File: day6/puzzle1.py
import re
grid = ([0] * 1000) * 1000
for i in open('day6.txt'):
 on = re.match(r"turn on ([0-9]+),([0-9]+) through ([0-9]+),([0-9]+)",i)
 if on:
  for y in xrange(int(on.group(2)), int(on.group(4)) + 1):
   for x in xrange(int(on.group(1)), int(on.group(3)) + 1):
    grid[x + y * 1000] = 1 
  continue
 off = re.match(r"turn off ([0-9]+),([0-9]+) through ([0-9]+),([0-9]+)",i)
 if off:
  for y in xrange(int(off.group(2)), int(off.group(4)) + 1):
   for x in xrange(int(off.group(1)), int(off.group(3)) + 1):
    grid[x + y * 1000] = 0 
  continue
 toggle = re.match(r"toggle ([0-9]+),([0-9]+) through ([0-9]+),([0-9]+)",i)
 if toggle:
  for y in xrange(int(toggle.group(2)), int(toggle.group(4)) + 1):
   for x in xrange(int(toggle.group(1)), int(toggle.group(3)) + 1):
    grid[x + y * 1000] = (grid[x + y * 1000] + 1) & 1 
  continue
print grid.count(1)

File: 2015/06/puzzle1.js
var fs = require('fs');

var ops = {
 'turn on': function () { return 1; },
 'turn off': function () { return 0; },
 'toggle': function (v) { return v^1; }
};

grid = new Array(1000 * 1000);

fs.readFile('day6.txt', 'utf-8', function (err, inp) {
 if (err) throw err;

 inp.trim().split(/\n/g).forEach(function (line) {
  var m = line.match(/(turn on|turn off|toggle) ([0-9]+),([0-9]+) through ([0-9]+),([0-9]+)/);
  if (m) {
   for (var y = Number(m[3]), yn = Number(m[5]); y <= yn; ++y) {
    for (var x = Number(m[2]), xn = Number(m[4]); x <= xn; ++x) {
     grid[y * 1000 + x] = ops[m[1]](grid[y * 1000 + x] || 0); 
    }
   }
  }
 });

 console.log(grid.filter(function (v) {
  return !!v;
 }).length);
});

